packages/lodestar/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -0,0 +1,310 @@
+      const participationStatus =
+        epoch === currentEpoch ? currentParticipation : epoch === currentEpoch - 1 ? previousParticipation : null;
+      if (!participationStatus) {
+        continue;
+      }
+      for (const attestationGroup of attestationGroups) {
+        const committee = attestationGroup.getCommittee();
+        const attestingIndices = committee.filter(
+          (index) => participationStatus[index] && participationStatus[index].timelySource
+        );
+        attestationGroup.removeBySeenValidators(attestingIndices);
+        attestations.push(...attestationGroup.getAttestations());
+      }
+    }
+    // consumer should limit MAX_ATTESTATIONS items
+    return isValidAttestation
+      ? attestations
+      : attestations.filter((attestation) => safeValidateAttestation(state, attestation));
+  }
+}
+
+function safeValidateAttestation(
+  state: CachedBeaconState<allForks.BeaconState>,
+  attestation: phase0.Attestation
+): boolean {
+  try {
+    phase0.validateAttestation(state, attestation);
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+interface AttestationWithIndex {
+  attestation: phase0.Attestation;
+  attestingIndices: ValidatorIndex[];
+  // this is <= attestingIndices.count since some attesters may be seen by the chain
+  // this is only updated and used in removeBySeenValidators function
+  freshAttesterCount?: number;
+}
+
+/**
+ * Maintain a pool of AggregatedAttestation which all share the same AttestationData.
+ * Preaggregate into smallest number of attestations.
+ * When getting attestations to be included in a block, sort by number of attesters.
+ * Use committee instead of aggregationBits to improve performance.
+ */
+export class MatchingDataAttestationGroup {
+  private readonly attestations: AttestationWithIndex[];
+  private readonly committee: ValidatorIndex[];
+  constructor(committee: ValidatorIndex[]) {
+    this.committee = committee;
+    this.attestations = [];
+  }
+
+  /** Add an attestation.
+   * Try to preaggregate to existing attestations if possible.
+   * If it's a subset of an existing attestations, it's not neccesrary to add to our pool.
+   */
+  add(attestation: AttestationWithIndex): InsertOutcome {
+    const attestingIndices = attestation.attestingIndices.valueOf() as ValidatorIndex[];
+    // preaggregate
+    let insertResult = InsertOutcome.NewData;
+    for (const existingAttestation of this.attestations) {
+      const existingAttestingIndices = existingAttestation.attestingIndices.valueOf() as ValidatorIndex[];
+      let numIntersection = 0;
+      for (const index of attestingIndices) {
+        if (existingAttestingIndices.includes(index)) numIntersection++;
+      }
+      // no intersection
+      if (numIntersection === 0) {
+        aggregateInto(existingAttestation, attestation, this.committee);
+        insertResult = InsertOutcome.Aggregated;
```

Comment:
Can we just return here
```ts
return InsertOutcome.Aggregated
```
?

Reply:
we want to aggregate all possible existing attestations so should not return there